src/relay/federation/mod.rs

```diff
@@ -0,0 +1,122 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use bb8_redis::{bb8::Pool, redis::cmd, RedisConnectionManager};
+use protobuf::Message;
+use workers::WorkerPool;
+
+use crate::types::Envelope;
+
+use self::router::Router;
+
+mod router;
+
+pub const DEFAULT_WORKERS: usize = 100;
+pub const FEDERATION_QUEUE: &str = "relay.federation";
+#[derive(Clone)]
+pub struct Federation {
+    redis_pool: Pool<RedisConnectionManager>,
+    workers: usize,
+}
+impl Federation {
+    pub fn new(redis_pool: Pool<RedisConnectionManager>) -> Self {
+        Self {
+            redis_pool,
+            workers: DEFAULT_WORKERS,
+        }
+    }
+}
+impl Federation {
+    pub fn with_workers(mut self, workers: usize) -> Self {
+        self.workers = workers;
+        self
+    }
+    // Sends federation msg to redis on (relay.federation)
+    pub async fn send<T: AsRef<[u8]>>(&self, msg: T) -> Result<()> {
+        let mut con = self.redis_pool.get().await?;
+        if let Err(err) = cmd("LPUSH")
+            .arg(FEDERATION_QUEUE)
+            .arg(msg.as_ref())
+            .query_async::<_, u32>(&mut *con)
+            .await
+        {
+            bail!("could not push msg to queue: {}", err)
+        };
+        Ok(())
+    }
+
+    // start polling from redis and send work to workers
+    pub async fn start(self) {
+        let work_runner = Router {};
+        let mut worker_pool = WorkerPool::new(Arc::new(work_runner), DEFAULT_WORKERS);
+        let mut con = match self.redis_pool.get().await {
+            Ok(con) => con,
+            Err(err) => {
+                panic!("could not get redis connection from pool, {}", err);
+            }
+        };
+        loop {
+            let worker_handler = worker_pool.get().await;
+            let (_, msg): (String, Vec<u8>) = match cmd("BRPOP")
+                .arg(FEDERATION_QUEUE)
+                .arg(0.0)
+                .query_async(&mut *con)
+                .await
+            {
+                Ok(msg) => msg,
+                Err(err) => {
+                    log::error!("could not get message from redis {}", err);
+                    continue;
+                }
```

Comment:
I think depending on this message, u either sleep for a second or two before u try again, or try to get a new connection from the pool in case connection was lost! 

May be it was also a bad idea to put the redis connection pool.get() outside the loop, lets assume that get ting a connection from the pool is cheap.